File: files/CH-166.R
library(tidyverse)
library(readxl)

path = "files/CH-166 Time Zone.xlsx"
input = read_excel(path, range = "B2:D12")
test  = read_excel(path, range = "F2:I12")

result = input %>%
  separate(`Date Time`, into = c("Day", "Month", "Year", "Hour", "Minute"), sep = " |:|/", convert = TRUE) %>%
  mutate(Day = Day + if_else(Hour >= 24, 1, 0),
         Hour = Hour %% 24) %>%
  mutate(`Date Time` = make_datetime(year = Year, month = Month, day = Day, hour = Hour, min = Minute),  .keep = "unused") %>%
  mutate(`GMT From` = str_remove(`GMT From`, "GMT") %>% as.numeric(),
         `GMT To` = str_remove(`GMT To`, "GMT") %>% as.numeric()) %>%
  mutate(`New Date Time` = `Date Time` + hours(`GMT To` - `GMT From`)) %>%
  select(`New Date Time`) %>%
  mutate(`New Date Time` = format(`New Date Time`, "%d/%m/%Y %H:%M"))
  

all.equal(result$`New Date Time`, test$`New Date Time`, check.attributes = FALSE)
# False, one discrepancy on 5th row.

File: files/100-199/166/CH-166.py
import pandas as pd

path = "CH-166 Time Zone.xlsx"
input = pd.read_excel(path, usecols="B:D", skiprows=1, nrows=10)
test = pd.read_excel(path, usecols="I:I", skiprows=1, nrows=10)

input[['Day', 'Month', 'Year', 'Hour', 'Minute']] = input['Date Time'].str.extract(r'(\d+)/(\d+)/(\d+) (\d+):(\d+)').astype(int)
input['Day'] += (input['Hour'] >= 24).astype(int)
input['Hour'] %= 24
input['Date Time'] = pd.to_datetime(input[['Year', 'Month', 'Day', 'Hour', 'Minute']])
input['GMT From'] = input['GMT From'].str.replace('GMT', '').astype(int)
input['GMT To'] = input['GMT To'].str.replace('GMT', '').astype(int)
input['New Date Time'] = input['Date Time'] + pd.to_timedelta(input['GMT To'] - input['GMT From'], unit='h')
input['New Date Time'] = input['New Date Time'].dt.strftime('%d/%m/%Y %H:%M')
result = input[['New Date Time']]

print(result == test) # False, Discrepancy on one row. Index 4.